### PR TITLE
feat(execute-operation): refactor codebase to isolate operation execution

### PR DESCRIPTION
**Pull request by mbroadst**
The gist of this change is that we were duplicating code _all over_ the codebase related to determining whether a callback or Promise was being used for most public facing API.  This PR addresses that, but also gains us the benefit of isolating the execution to a single point such that we can now introduce operation-wide changes very easily - most importantly the requirement set forth in the Driver Sessions spec that we use an implicit session if none is provided for _all_ public operations.

**NOTE:** This is rebased upon the `bulk-write-error` branch, so you might see some cross-pollination with the commits until that branch is merged (hopefully tomorrow).

### CODE REVIEW

**Comment by kmahar** [comment]
man I can't believe how much copy paste "// Execute using callback... // Return a Promise" there was before. this is much nicer

File: lib/collection.js [hunk 1]
```diff
@@ -1875,22 +1690,10 @@ define.classMethod('dropIndex', { callback: true, promise: true });
  * @return {Promise} returns Promise if no callback passed
  */
 Collection.prototype.dropIndexes = function(options, callback) {
-  var self = this;
-
-  // Do we have options
```
Comment (kmahar) [inline lib/collection.js]: do we? the world may never know

File: lib/collection.js [hunk 2]
```diff
@@ -2999,25 +2664,19 @@ define.classMethod('parallelCollectionScan', { callback: true, promise: true });
  * @return {Promise} returns Promise if no callback passed
  */
 Collection.prototype.geoNear = function(x, y, options, callback) {
-  var self = this;
   var point = typeof x === 'object' && x,
     args = Array.prototype.slice.call(arguments, point ? 1 : 2);
 
   callback = args.pop();
```
Comment (kmahar) [inline lib/collection.js]: if you feel like removing 4 more lines of code from each function, I think that just doing

`callback = typeof args[args.length-1] === 'function' ? args.pop() : undefined`

would accomplish the same thing? kinda weird to remove the last element and then sometimes put it back. (I know you didn't introduce that though so feel free to ignore this)

File: lib/gridfs-stream/index.js [hunk 3]
```diff
@@ -295,20 +283,7 @@ GridFSBucket.prototype.openDownloadStreamByName = function(filename, options) {
  */
 
 GridFSBucket.prototype.rename = function(id, filename, callback) {
-  if (typeof callback === 'function') {
-    return _rename(this, id, filename, callback);
-  }
-
-  var _this = this;
```
Comment (kmahar) [inline lib/gridfs-stream/index.js]: just curious, what would the point of doing this have been?

Comment (kmahar) [inline lib/gridfs-stream/index.js]: just curious, what would the point of doing this have been?

File: lib/mongo_client.js [hunk 4]
```diff
@@ -217,31 +221,22 @@ var define = (MongoClient.define = new Define('MongoClient', MongoClient, false)
...
-      // Attempt to connect
-      connect(self, self.s.url, self.s.options, function(err, topology) {
-        if (err) return reject(err);
-        self.topology = topology;
-        resolve(self);
-      });
-    });
+  var err = validOptions(this.s.options);
+
+  if (typeof callback === 'string') {
```
Comment (kmahar) [inline lib/mongo_client.js]: why specifically check for a string?

File: lib/collection.js (hunk 2 above)
Comment (kmahar) [reply lib/collection.js]: oh just realized we already do this in MongoClient.connect lol

Comment (mbroadst) [reply lib/collection.js]: Good call though! I went through to refactor this and found a number of places I had missed migration to `executeOperation` in GridFS code in the process 💃

File: lib/collection.js (hunk 1 above)
Comment (mbroadst) [reply lib/collection.js]: lol

File: lib/gridfs-stream/index.js (hunk 3 above)
Comment (mbroadst) [reply lib/gridfs-stream/index.js]: using `_this`? It's the same idea as `var self = this;` you're just using a locally scoped version of `this` for use within a closure.

File: lib/mongo_client.js (hunk 4 above)
Comment (mbroadst) [reply lib/mongo_client.js]: honestly, I didn't dig into it but left it there for legacy reasons. This was a big enough refactoring that it seemed important to reduce the variables in terms of code changes.

Comment (kmahar) [reply lib/mongo_client.js]: oh oops, was this in another file before or something? (highlighted in green in the diff so I thought new)

File: lib/db.js [hunk 5]
```diff
@@ -1093,26 +1015,15 @@ define.classMethod('executeDbAdminCommand', { callback: true, promise: true });
...
-  callback = args.pop();
-  if (typeof callback !== 'function') args.push(callback);
+  callback = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
   options = args.length ? args.shift() || {} : {};
   options = typeof callback === 'function' ? options : callback;
   options = options == null ? {} : options;
   // Shallow clone the options
   options = shallowClone(options);
 
-  // If we have a callback fallback
```
Comment (kmahar) [inline lib/db.js]: poetry

**Comment by kmahar** [comment]
ok done commenting

**Pull request merged**